### DISCUSSION TITLE
Triangulate MultiSurface

### DISCUSSION THREAD

**Discussion by d1kkop** [cityjson/specs#162] [Q&A] [answered]
Hello, 

Should a multisurface be triangulated using earcut algorithm? So, an outerior ring and possible multiple interior rings?
If so, then how to do this since it are 3d coordinates?

Or if not, how should it be triangulated? I am using C#.

Kind regards,

Bart

**Comment by hugoledoux** [comment]
MultiSurfaces in CityJSON do no need to be triangulated (we use ISO 19107, which is [summarised here](https://val3dity.readthedocs.io/en/latest/definitions/#iso19107-primitives)), but they can.

If you want to triangulate your surfaces, you can use [cjio](https://github.com/cityjson/cjio) with the operator `triangulate`

**Reply by d1kkop** [reply]
Hi hugoledoux.

Thanks for reaching. However, in a game engine and ultimately DirectxX, Vulkan or (Web)GL, it must be converted to triangles or quads.
I am also little confused, bacause the drawing here: https://www.cityjson.org/dev/geom-arrays/ shows that a multisurface can contains holes. 
I try to impelemnt this in Unity 3D engine, but converting it in python complicates the workflow a lot.  Unity works with C# and Unreal with C++.

Kind regards,
Bart

**Comment by kenohori** [comment]
You basically need to project each multisurface to 2D, triangulate, and then go back to 3D. The earcut algorithm should work as long as it supports holes, like in the Mapbox implementation. In C++, you could use the constrained triangulations in CGAL or Triangle. No idea about C#.

**Reply by d1kkop** [reply]
I thought so too. But it got me confused a bit if this is correct. For instance for vertical surfaces (walls), it cannot be projected  to the ground, so special casing is necessary. 
I wonder if the following would yield correct results (assuming Z is up).
1. Obtain normal from surface.
2. if (abs(normal.Z) < 0.5) project to the side (X, or Y), else project to ground. 
3. Do earcut
4. Project back

I thnk it is, vertices are not adjusted.

Edit: might be wise to do additional check to see if X or Y normal is not almost 1, and based on that project either way.

**Comment by hugoledoux** [comment, marked as answer]
I am not entirely sure I understand the question, but if it's "how to triangulate a MultiSurface" then yes mapbox-earcut works for polygons with holes, but I found it to be producing collapsed triangles (area=0.0; or 3 points collinear). For Unity I assume this is not a problem though.

As @kenohori pointed out, CGAL is perhaps a better option, but no C#...

**Reply by d1kkop** [reply]
Thank you for the answer. Together with answer from knohori, I think I have a solution.

**Comment by d1kkop** [comment]
Here unity implementation :-)
Dont fotget to put on HD.
https://www.youtube.com/watch?v=4a5TVG-6gck&ab_channel=BartKnuiman

**Comment by kenohori** [comment]
very nice!
